drizzle/resample.py

```diff
@@ -0,0 +1,404 @@
+            raise ValueError(
+                "'pixmap' shape is not consistent with 'data' shape."
+            )
+
+        # TODO: this is code that would enable initializer to not need output
+        #       image shape at all and set output image shape based on
+        #       output coordinates from the pixmap. If this is enabled, then
+        #       some checks in __init__ may be removed:
+        # if self.out_shape is None:
+        #     pmap_xmin = int(np.floor(np.min(pixmap[:, :, 0])))
+        #     pmap_xmax = int(np.ceil(np.max(pixmap[:, :, 0])))
+        #     pmap_ymin = int(np.floor(np.min(pixmap[:, :, 1])))
+        #     pmap_ymax = int(np.ceil(np.max(pixmap[:, :, 1])))
+        #     pixmap = pixmap.copy()
+        #     pixmap[:, :, 0] -= pmap_xmin
+        #     pixmap[:, :, 1] -= pmap_ymin
+        #     self.out_shape = (
+        #         pmap_xmax - pmap_xmin + 1,
+        #         pmap_ymax - pmap_ymin + 1
+        #     )
+
+        if xmin is None or xmin < 0:
+            xmin = 0
+
+        if ymin is None or ymin < 0:
+            ymin = 0
+
+        if xmax is None or xmax > self.out_shape[1] - 1:
+            xmax = self.out_shape[1] - 1
+
+        if ymax is None or ymax > self.out_shape[0] - 1:
+            ymax = self.out_shape[0] - 1
+
+        if weight_map is not None:
+            weight_map = np.asarray(weight_map, dtype=np.float32)
+        else:  # TODO: this should not be needed after C code modifications
+            weight_map = np.ones_like(data)
+
+        pixmap = np.asarray(pixmap, dtype=np.float64)
+
+        # TODO: probably tdriz should be modified to not return version.
+        #       we should not have git, Python, C, ... versions
+
+        # TODO: While drizzle code in cdrizzlebox.c supports weight_map=None,
+        #       cdrizzleapi.c does not. It should be modified to support this
+        #       for performance reasons.
+        if len(self.out_ctx.shape) == 2:
+            assert plane_no == 0
+            ctx_plane = self.out_ctx
+        else:
+            ctx_plane = self.out_ctx[plane_no]
+
+        _vers, nmiss, nskip = cdrizzle.tdriz(
+            data,
+            weight_map,
+            pixmap,
+            self.out_sci,
+            self.out_wht,
+            ctx_plane,
+            uniqid=id_in_plane + 1,
+            xmin=xmin,
+            xmax=xmax,
+            ymin=ymin,
+            ymax=ymax,
+            scale=scale, # scales image intensity. usually equal to pixel scale
+            pixfrac=pixfrac,
+            kernel=self.kernel,
+            in_units=in_units,
+            expscale=expscale,
+            wtscale=wht_scale,
+            fillstr=self.fillval
+        )
+        self._cversion = _vers  # TODO: probably not needed
+
+        return nmiss, nskip
+
+
+def blot_image(data, pixmap, pix_ratio, exptime, output_pixel_shape,
+               interp='poly5', sinscl=1.0):
+    """
+    Resample input image using interpolation to an output grid.
```

Comment:
I find this confusing in that the term input seems ambiguous. Is that the input of the original image, or the input of this function. I would start saying that this is intended to do the inverse resampling that drizzle does.

Reply:
these are comments from the existing code. I did not modify them at this moment. I only moved the `blot_image` out of the `Drizzle` class as it does not belong to drizzle logically and does not need to be in a class.

I do not know who came up with the term `blot` but it is definitely not me.

> this is intended to do the inverse resampling that drizzle does.

I disagree. I think there is nothing "inverse" in here. Resampling is just resampling no matter the direction. Drizzle mixes values from several input pixels into one output pixel. The term "inverse" would suggest un-mingling those values to get the input (to drizzle) image, like Lucy-Richardson deconvolution.

Reply:
Well, I think that an inverse mapping is tightly associated with the term "blot" (I'm not sure why it is called that either). Perhaps a round tripping is the right term, where it is understood that it is not a pure inverse. So I think my comment stands to make the issue clearer since those using blot are almost always using the inverse mapping, aren't they? If you got rid of `blot` and forced people to use drizzling directly, then this wouldn't be an issue. But there is a history to this term.

Reply:
Resampling algorithms (interpolating, like "blot", or drizzle) are all "one way" in the sense that you cannot apply it twice with "inverse mapping" (i.e., inverted pixmap) and expect things to "round-trip". Round-tripping will work with "blot" but only if image features can be perfectly (locally) modeled by linear/poly3/poly5 functions. With these kinds of images, in principle, we could use blot to resample from one frame to another and back. Reverse transformation in this case would be performed by _exactly the same_ "blot" algorithm (and code). The only difference is in pixmap. Pixmap is always from the original (image's) coordinate system to the output (destination) coordinate system. 

In any case,
```
I would start saying that this is intended to do the inverse resampling that drizzle does.
```
IMO is fundamentally incorrect: one cannot do "inverse resampling" of drizzled images using interpolation (possibly except for point kernel in the absence of distortions, etc.)

Reply:
"inverse resampling" suggests inverse operation to drizzle which is not what blot is doing. "blot" can be used both for forward (meaning resampling a distorted image onto an undistorted frame) transform and backward transform (from the distortion-corrected frame to the distorted frame). We usually use it for the "backward" transformation but it is not the same as "inverse": the final image is not the initial image used as input to drizzle.

Reply:
I think we are talking at cross-purposes. What I mean is that blot is usually taken as mapping in the opposite direction to the original sampling. It is not an inverse of the original resampling, for sure. The point I was making was that there is ambiguity as to what input means, i.e., the original input image used for the original resampling, or the product in the destination pixel coordinate system to be mapped back to the original pixel coordinate system. In other words, with blot one could be confused as to which array is being referred to. It would be nice to clarify that in this case the input is the image generated from the original resampling (or whatever words describe that best). Maybe the current wording is obvious to everyone else, but it wasn't for me. Perhaps we should take a vote :-)